src/rpc/mempool.cpp

```diff
@@ -888,34 +890,57 @@ static RPCHelpMan submitpackage()
                 {
-                    throw JSONRPCTransactionError(TransactionError::INVALID_PACKAGE,
-                        package_result.m_state.GetRejectReason());
+                    // Everything in the package should be submitted
+                    CHECK_NONFATAL(package_result.m_tx_results.size() == txns.size());
+                    for (const auto& tx : txns) {
+                        wtixd_to_error_strings.insert({tx->GetWitnessHash(), std::nullopt});
+                    }
+                    break;
                 }
                 case PackageValidationResult::PCKG_MEMPOOL_ERROR:
                 {
+                    // This only happens with internal bug; user should stop and report
                     throw JSONRPCTransactionError(TransactionError::MEMPOOL_ERROR,
                         package_result.m_state.GetRejectReason());
                 }
+                case PackageValidationResult::PCKG_POLICY:
                 case PackageValidationResult::PCKG_TX:
                 {
+                    // Package-wide error we want to return, but we also want to return individual responses
+                    package_msg = package_result.m_state.GetRejectReason();
+
                     for (const auto& tx : txns) {
                         auto it = package_result.m_tx_results.find(tx->GetWitnessHash());
-                        if (it != package_result.m_tx_results.end() && it->second.m_state.IsInvalid()) {
-                            throw JSONRPCTransactionError(TransactionError::MEMPOOL_REJECTED,
-                                strprintf("%s failed: %s", tx->GetHash().ToString(), it->second.m_state.GetRejectReason()));
+                        if (it != package_result.m_tx_results.end()) {
+                            if (it->second.m_state.IsInvalid()) {
+                                wtixd_to_error_strings.insert({tx->GetWitnessHash(), it->second.m_state.ToString()});
+                            } else {
+                                wtixd_to_error_strings.insert({tx->GetWitnessHash(), std::nullopt});
+                            }
+                        } else {
+                            // No result returned; report this helpfully
+                            wtixd_to_error_strings.insert({tx->GetWitnessHash(), "unevaluated"});
                         }
                     }
-                    // If a PCKG_TX error was returned, there must have been an invalid transaction.
-                    NONFATAL_UNREACHABLE();
+                    break;
                 }
             }
+
             size_t num_broadcast{0};
             for (const auto& tx : txns) {
+                // We don't want to re-submit the txn for validation in BroadcastTransaction
+                if (!mempool.exists(GenTxid::Txid(tx->GetHash()))) {
+                    continue;
+                }
+
+                // We do not expect an error here; we are only broadcasting things already/still in mempool
```

Comment:
"all transactions were submitted" is not necessarily true anymore, no?

Reply:
fixed